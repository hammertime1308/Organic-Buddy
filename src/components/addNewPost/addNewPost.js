import React, { useState } from 'react';
import {
  ActivityIndicator,
  Button,
  Image,
  Modal,
  ScrollView,
  Text,
  TextInput,
  TouchableOpacity,
} from 'react-native';
import { launchImageLibrary } from 'react-native-image-picker';
import styled from 'styled-components/native';

import { newPost } from '../../api';

import { COLOR } from '../../styles';

const ModalContainer = styled.View`
  flex: 1;
  background-color: rgba(0, 0, 0, 0.5);
  justify-content: center;
  align-items: center;
`;

const ModalView = styled.View`
  background-color: ${COLOR.White};
  width: 90%;
  height: 500px;
  border-radius: 5px;
  align-items: center;
`;

export const AddNewPost = ({ visible, user, closeModal }) => {
  const [title, setTitle] = useState('');
  const [description, setDescription] = useState('');
  const [image, setImage] = useState('');

  const [loading, setLoading] = useState(false);

  const clearData = () => {
    setTitle('');
    setDescription('');
    setImage('');
  };

  const isValid = () => {
    if (title === '') {
      alert('Add Post title');
      return false;
    } else if (description === '') {
      alert('Add Post description');
      return false;
    }
    return true;
  };

  const selectImages = async () => {
    let options = {
      title: 'Select Image',
      mediaType: 'photo',
      quality: 0.3,
      includeBase64: true,
      storageOptions: {
        skipBackup: true,
        path: 'images',
      },
    };
    launchImageLibrary(options, response => {
      if (response.didCancel) {
        setImage('');
      } else if (response.error) {
        alert('Error: ', response.error);
      } else {
        setImage(response.base64);
      }
    });
  };

  const apiCall = async () => {
    if (isValid()) {
      setLoading(true);
      let response = await newPost(
        user.userId,
        title,
        description,
        image === '' ? [] : [image],
      );
      setLoading(false);
      if (response.status === 200) {
        clearData();
        closeModal();
      } else {
        alert(response.data);
      }
    }
  };

  return (
    <Modal
      visible={visible}
      transparent
      onRequestClose={() => {
        clearData();
        closeModal();
      }}
      animationType="fade">
      <ModalContainer>
        <ModalView>
          <Text
            style={{
              fontSize: 20,
              fontWeight: 'bold',
              color: `${COLOR.GreenHeader}`,
              paddingTop: 20,
            }}>
            Create a new post
          </Text>
          {loading ? (
            <ActivityIndicator
              style={{
                alignItems: 'center',
                height: 400,
                width: '100%',
              }}
              size={50}
              color={COLOR.GreenHeader}
            />
          ) : (
            <ScrollView
              style={{
                flex: 1,
                width: '100%',
                padding: 20,
              }}>
              <TextInput
                placeholder="Title for post (3-4 words)"
                value={title}
                onChangeText={setTitle}
                style={{
                  borderColor: 'black',
                  borderWidth: 0.2,
                  textAlign: 'center',
                  marginTop: 20,
                }}
              />
              <TextInput
                placeholder="Description of post (explain the issue)"
                value={description}
                onChangeText={setDescription}
                style={{
                  borderColor: 'black',
                  borderWidth: 0.2,
                  textAlign: 'center',
                  marginTop: 20,
                  marginBottom: 20,
                }}
              />

              <Image
                style={{
                  width: 100,
                  height: 100,
                  marginBottom: 20,
                  alignSelf: 'center',
                }}
                source={{
                  uri: 'data:image/jpeg;base64,' + image,
                }}
              />

              <Button
                title="add image"
                color={COLOR.GreenHeader}
                onPress={selectImages}
              />
            </ScrollView>
          )}
          <TouchableOpacity
            activeOpacity={0.8}
            onPress={apiCall}
            style={{
              backgroundColor: `${COLOR.GreenHeader}`,
              position: 'absolute',
              bottom: 0,
              width: '100%',
              height: '10%',
              justifyContent: 'center',
              alignItems: 'center',
              borderBottomLeftRadius: 5,
              borderBottomRightRadius: 5,
            }}>
            <Text style={{ fontSize: 20, color: 'white', fontWeight: 'bold' }}>
              Post
            </Text>
          </TouchableOpacity>
        </ModalView>
      </ModalContainer>
    </Modal>
  );
};
