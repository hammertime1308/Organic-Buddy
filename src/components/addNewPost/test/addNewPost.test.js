import React from 'react';
import renderer from 'react-test-renderer';
import { AddNewPost } from '../addNewPost';

describe('Add new post', () => {
  it('Should not render Modal', () => {
    const tree = renderer
      .create(
        <AddNewPost
          visible={false}
          user={{ userId: '0' }}
          closeModal={() => {}}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should render modal', () => {
    const tree = renderer
      .create(
        <AddNewPost
          visible={true}
          user={{ userId: '0' }}
          closeModal={() => {}}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
