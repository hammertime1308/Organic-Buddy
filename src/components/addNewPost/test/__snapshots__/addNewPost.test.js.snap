// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add new post Should not render Modal 1`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  transparent={true}
  visible={false}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(0, 0, 0, 0.5)",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "justifyContent": "center",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffff",
            "borderRadius": 5,
            "height": 500,
            "width": "90%",
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "rgba(76, 154, 42, 1)",
            "fontSize": 20,
            "fontWeight": "bold",
            "paddingTop": 20,
          }
        }
      >
        Create a new post
      </Text>
      <RCTScrollView
        style={
          Object {
            "flex": 1,
            "padding": 20,
            "width": "100%",
          }
        }
      >
        <View>
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            placeholder="Title for post (3-4 words)"
            rejectResponderTermination={true}
            style={
              Object {
                "borderColor": "black",
                "borderWidth": 0.2,
                "marginTop": 20,
                "textAlign": "center",
              }
            }
            underlineColorAndroid="transparent"
            value=""
          />
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            placeholder="Description of post (explain the issue)"
            rejectResponderTermination={true}
            style={
              Object {
                "borderColor": "black",
                "borderWidth": 0.2,
                "marginBottom": 20,
                "marginTop": 20,
                "textAlign": "center",
              }
            }
            underlineColorAndroid="transparent"
            value=""
          />
          <Image
            source={
              Object {
                "uri": "data:image/jpeg;base64,",
              }
            }
            style={
              Object {
                "alignSelf": "center",
                "height": 100,
                "marginBottom": 20,
                "width": 100,
              }
            }
          />
          <View
            accessibilityRole="button"
            accessibilityState={Object {}}
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {},
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "margin": 8,
                      "textAlign": "center",
                    },
                    Object {
                      "color": "rgba(76, 154, 42, 1)",
                    },
                  ]
                }
              >
                add image
              </Text>
            </View>
          </View>
        </View>
      </RCTScrollView>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(76, 154, 42, 1)",
            "borderBottomLeftRadius": 5,
            "borderBottomRightRadius": 5,
            "bottom": 0,
            "height": "10%",
            "justifyContent": "center",
            "opacity": 1,
            "position": "absolute",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 20,
              "fontWeight": "bold",
            }
          }
        >
          Post
        </Text>
      </View>
    </View>
  </View>
</Modal>
`;

exports[`Add new post Should render modal 1`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  transparent={true}
  visible={true}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(0, 0, 0, 0.5)",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "justifyContent": "center",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffff",
            "borderRadius": 5,
            "height": 500,
            "width": "90%",
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "rgba(76, 154, 42, 1)",
            "fontSize": 20,
            "fontWeight": "bold",
            "paddingTop": 20,
          }
        }
      >
        Create a new post
      </Text>
      <RCTScrollView
        style={
          Object {
            "flex": 1,
            "padding": 20,
            "width": "100%",
          }
        }
      >
        <View>
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            placeholder="Title for post (3-4 words)"
            rejectResponderTermination={true}
            style={
              Object {
                "borderColor": "black",
                "borderWidth": 0.2,
                "marginTop": 20,
                "textAlign": "center",
              }
            }
            underlineColorAndroid="transparent"
            value=""
          />
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            placeholder="Description of post (explain the issue)"
            rejectResponderTermination={true}
            style={
              Object {
                "borderColor": "black",
                "borderWidth": 0.2,
                "marginBottom": 20,
                "marginTop": 20,
                "textAlign": "center",
              }
            }
            underlineColorAndroid="transparent"
            value=""
          />
          <Image
            source={
              Object {
                "uri": "data:image/jpeg;base64,",
              }
            }
            style={
              Object {
                "alignSelf": "center",
                "height": 100,
                "marginBottom": 20,
                "width": 100,
              }
            }
          />
          <View
            accessibilityRole="button"
            accessibilityState={Object {}}
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {},
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "margin": 8,
                      "textAlign": "center",
                    },
                    Object {
                      "color": "rgba(76, 154, 42, 1)",
                    },
                  ]
                }
              >
                add image
              </Text>
            </View>
          </View>
        </View>
      </RCTScrollView>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(76, 154, 42, 1)",
            "borderBottomLeftRadius": 5,
            "borderBottomRightRadius": 5,
            "bottom": 0,
            "height": "10%",
            "justifyContent": "center",
            "opacity": 1,
            "position": "absolute",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 20,
              "fontWeight": "bold",
            }
          }
        >
          Post
        </Text>
      </View>
    </View>
  </View>
</Modal>
`;
